internal/provider/resource_access_offer.go

```diff
@@ -0,0 +1,547 @@
+		if resp.Diagnostics.HasError() {
+			return
+		}
+		state.ConsumeUsers = consumeUsersSet
+	}
+	// Save read users to state
+	if len(users[permission.ReadAccess]) > 0 {
+		readUsersSet, errDiag := basetypes.NewSetValueFrom(ctx, types.StringType, users[permission.ReadAccess])
+		resp.Diagnostics.Append(errDiag...)
+		if resp.Diagnostics.HasError() {
+			return
+		}
+		state.ReadUsers = readUsersSet
+	}
+	// Set the plan onto the Terraform state
+	state.OfferURL = types.StringValue(offerURL)
+	resp.Diagnostics.Append(resp.State.Set(ctx, &state)...)
+}
+
+// Update attempts to update the access to the offer.
+func (a *accessOfferResource) Update(ctx context.Context, req resource.UpdateRequest, resp *resource.UpdateResponse) {
+	// Check first if the client is configured
+	if a.client == nil {
+		addClientNotConfiguredError(&resp.Diagnostics, "access offer", "update")
+		return
+	}
+	var plan, state accessOfferResourceOffer
+
+	// Read Terraform configuration from the request into the model
+	resp.Diagnostics.Append(req.Plan.Get(ctx, &plan)...)
+	resp.Diagnostics.Append(req.State.Get(ctx, &state)...)
+	if resp.Diagnostics.HasError() {
+		return
+	}
+
+	// Get the users to grant admin
+	var adminUsers []string
+	if !plan.AdminUsers.IsNull() {
+		resp.Diagnostics.Append(plan.AdminUsers.ElementsAs(ctx, &adminUsers, false)...)
+		if resp.Diagnostics.HasError() {
+			return
+		}
+	}
+
+	// Get the users to grant consume
+	var consumeUsers []string
+	if !plan.ConsumeUsers.IsNull() {
+		resp.Diagnostics.Append(plan.ConsumeUsers.ElementsAs(ctx, &consumeUsers, false)...)
+		if resp.Diagnostics.HasError() {
+			return
+		}
+	}
+
+	// Get the users to grant read
+	var readUsers []string
+	if !plan.ReadUsers.IsNull() {
+		resp.Diagnostics.Append(plan.ReadUsers.ElementsAs(ctx, &readUsers, false)...)
+		if resp.Diagnostics.HasError() {
+			return
+		}
+	}
+
+	// validate if there are overlaps or admin user
+	// validation is done here considering dynamic (juju_user resource) and static values for users
+	err := validateNoOverlapsNoAdmin(adminUsers, consumeUsers, readUsers)
+	if err != nil {
+		resp.Diagnostics.AddError("Client Error", fmt.Sprintf("Unable to update access offer resource, got error: %s", err))
+		return
+	}
+
+	// Get users from state
+	var readStateUsers []string
+	if !state.ReadUsers.IsNull() {
+		resp.Diagnostics.Append(state.ReadUsers.ElementsAs(ctx, &readStateUsers, false)...)
+		if resp.Diagnostics.HasError() {
+			return
+		}
+	}
+
+	var consumeStateUsers []string
+	if !state.ConsumeUsers.IsNull() {
+		resp.Diagnostics.Append(state.ConsumeUsers.ElementsAs(ctx, &consumeStateUsers, false)...)
+		if resp.Diagnostics.HasError() {
+			return
+		}
+	}
+
+	var adminStateUsers []string
+	if !state.AdminUsers.IsNull() {
+		resp.Diagnostics.Append(state.AdminUsers.ElementsAs(ctx, &adminStateUsers, false)...)
+		if resp.Diagnostics.HasError() {
+			return
+		}
+	}
+
+	// Revoke read (remove access) of all state users
```

Comment:
Why revoke access from all users and not just those users whose access has been removed?
And then add access for new users.
Is it just for simplicity sake?

Reply:
Discussed here :)

https://github.com/juju/terraform-provider-juju/pull/633#discussion_r1907215101

Reply:
Gotcha, ty. I'd suggest adding a comment explaining thinking. Coming across this without explanation I'd be surprised.